design/src/server/anatomy.md

```diff
@@ -0,0 +1,843 @@
+
+```rust
+/// Provides a standard interface for applying [`Plugin`]s to a service builder. This is implemented automatically for
+/// all builders.
+///
+/// As [`Plugin`]s modify the way in which [`Operation`]s are [`upgraded`](crate::operation::Upgradable) we can use
+/// [`Pluggable`] as a foundation to write extension traits which are implemented for all service builders.
+///
+/// # Example
+///
+/// ```
+/// # struct PrintPlugin;
+/// # use aws_smithy_http_server::plugin::Pluggable;
+/// trait PrintExt: Pluggable<PrintPlugin> {
+///     fn print(self) -> Self::Output where Self: Sized {
+///         self.apply(PrintPlugin)
+///     }
+/// }
+///
+/// impl<Builder> PrintExt for Builder where Builder: Pluggable<PrintPlugin> {}
+/// ```
+pub trait Pluggable<NewPlugin> {
+    type Output;
+
+    /// Applies a [`Plugin`] to the service builder.
+    fn apply(self, plugin: NewPlugin) -> Self::Output;
+}
+```
+
+As seen in the `Pluggable` documentation, third-parties can use extension traits over `Pluggable` to extend the API of builders. In addition to all the `Op{N}` the service builder also holds a `Pl`:
+
+```rust
+/// The service builder for [`PokemonService`].
+///
+/// Constructed via [`PokemonService::builder`].
+pub struct PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6, Pl> {
+    capture_pokemon_operation: Op1,
+    empty_operation: Op2,
+    get_pokemon_species: Op3,
+    get_server_statistics: Op4,
+    get_storage: Op5,
+    health_check_operation: Op6,
+
+    plugin: Pl
+}
+```
+
+which allows the following `Pluggable` implementation to be generated:
+
+```rust
+impl<Op1, Op2, /* ... */, Pl, NewPl> Pluggable<NewPl> for PokemonServiceBuilder<Op1, Op2, /* ... */, Pl>
+{
+    type Output = PokemonServiceBuilder<Op1, Exts1, PluginStack<Pl, NewPl>>;
+    fn apply(self, plugin: NewPl) -> Self::Output {
+        PokemonServiceBuilder {
+            capture_pokemon_operation: self.capture_pokemon_operation,
+            empty_operation: self.empty_operation,
+            /* ... */,
+
+            plugin: PluginStack::new(self.plugin, plugin),
+        }
+    }
+}
+```
+
+Here `PluginStack` works in a similar way to [`tower::layer::util::Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html) - allowing users to stack a new plugin rather than replacing the currently set one.
+
+## Accessing Unmodelled Data
+
+An additional omitted detail is that we provide an "escape hatch" allowing `Handler`s and `OperationService`s to accept data that isn't modelled. In addition to accepting `Op::Input` they can accept additional arguments which implement the [`FromParts`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/request.rs#L114-L121) trait:
```

Comment:
Do we have any example of using `FromParts`? If so, it would probably be useful to link it here.